File: py/2022/day02.py
from aoc import main
from aoc.util import mod_range

Game = tuple[int, int]


def parse(s: str) -> list[Game]:
    return [
        ("ABC".index(a) + 1, "XYZ".index(b) + 1)
        for a, b in [line.split(" ", 2) for line in s.splitlines()]
    ]


def score(a: int, b: int) -> int:
    score = b
    if a == b:
        score += 3
    elif mod_range(a + 1, 1, 4) == b:
        score += 6
    return score


def strategize(game: Game) -> Game:
    opp, strategy = game
    DELTAS = {1: -1, 2: 0, 3: 1}
    return opp, mod_range(opp + DELTAS[strategy], 1, 4)


if __name__ == "__main__":
    main(
        lambda s: sum(score(a, b) for a, b in parse(s)),
        lambda s: sum(score(*strategize(game)) for game in parse(s)),
    )

File: ts/2022/day02.ts
import { main } from 'lib/advent';
import { Iter, iter } from 'lib/iter';
import { ValuesOf, lines } from 'lib/util';

type Value = 1 | 2 | 3;
function parse(lines: string[]): Iter<Value[]> {
  const map: Record<string, Value> = { A: 1, B: 2, C: 3, X: 1, Y: 2, Z: 3 };
  return iter(lines).map((line) => line.split(' ', 2).map((v) => map[v]));
}
const winsAgainst = (v: Value): Value => ((v % 3) + 1) as Value;
const losesAgainst = (v: Value): Value => (((v + 1) % 3) + 1) as Value;

const Strategy = {
  Lose: 1,
  Draw: 2,
  Win: 3,
} as const;
type Strategy = ValuesOf<typeof Strategy>;

function score([a, b]: Value[]): number {
  let score = b;
  if (a === b) score += 3;
  else if (b === winsAgainst(a)) score += 6;
  return score;
}

function strategizedScore([a, b]: Value[]): number {
  if (b === Strategy.Lose) b = losesAgainst(a);
  else if (b === Strategy.Draw) b = a;
  else if (b === Strategy.Win) b = winsAgainst(a);
  return score([a, b]);
}

main(
  (s) => parse(lines(s)).map(score).sum(),
  (s) => parse(lines(s)).map(strategizedScore).sum()
);
